### PR TITLE
 [core][autoscaler][v2] Subscribers - cloud instances updater [1/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR refactors the `InstanceLauncher` subscriber into a `CloudInstanceUpdater` which acts as a proxy to make calls to the cloud provider for launching and terminating cloud instances:
- It also removes the dependencies to `InstanceStorage` since the subscriber now should not be updating instance storage, the `Reconciler` is expected to be the only place of status transitions. 

This works closely with https://github.com/ray-project/ray/pull/42973

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/subscribers/instance_launcher.py [hunk 1]
```diff
@@ -1,160 +0,0 @@
...
-    def _may_launch_new_instances(self):
-        new_instances, _ = self._instance_storage.get_instances(
-            status_filter={Instance.UNKNOWN}
-        )
-
-        if not new_instances:
-            logger.debug("No instances to launch")
-            return
-
-        queued_instances = []
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/subscribers/instance_launcher.py:59]: the logic here for converting QUEUED to REQUESTED has been now moved to this PR https://github.com/ray-project/ray/pull/42973/files

File: python/ray/autoscaler/v2/instance_manager/subscribers/cloud_instance_updater.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    def notify(self, events: List[InstanceUpdateEvent]) -> None:
+        new_requests = [
+            event for event in events if event.new_instance_status == Instance.REQUESTED
+        ]
+        new_terminations = [
+            event
+            for event in events
+            if event.new_instance_status == Instance.TERMINATING
+        ]
+        self._executor.submit(self._launch_new_instances, new_requests)
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/subscribers/cloud_instance_updater.py:43]: why do we need to run inside executor if node provider api is async?

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/subscribers/cloud_instance_updater.py:43]: good point - we could just run w/o executor now.

**Comment by iycheng** [comment]
The PR looks good to me. I'm wondering how is the error being handled?

**Comment by rickyyx** [comment]
> The PR looks good to me. I'm wondering how is the error being handled?

Errors will be made available through cloud provider's `poll_errors()`. The cloud provider is now async, and would push errors to a queue internally that can be polled by callers.  See this

https://github.com/ray-project/ray/blob/6b6f23a370d8e78c709f36d51014083ead9a2447/python/ray/autoscaler/v2/instance_manager/node_provider.py#L76-L137

File: python/ray/autoscaler/v2/instance_manager/subscribers/cloud_instance_updater.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        if not new_terminations:
+            logger.debug("No instances to terminate.")
+            return
+
+        # Terminate the instances.
+        cloud_instance_ids = [event.cloud_instance_id for event in new_terminations]
+
+        # This is an async call.
+        self._cloud_provider.terminate(
+            ids=cloud_instance_ids, request_id=str(time.time_ns())
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/subscribers/cloud_instance_updater.py:62]: Should we use uuid?

**Review by jjyao — APPROVED** [review]

**Pull request merged**